### PR TITLE
python3Packages.catboost: 0.24.4 -> 1.0.5

### PR DESCRIPTION
**Pull request by pbsds**
###### Description of changes

Shit was [broken](https://hydra.nixos.org/build/179808466), so I upgraded it.
We're now on a stable release, breaking changes include cuda 11 being default.

https://github.com/catboost/catboost/releases

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by onny** [comment]
@ofburg build

**Comment by pbsds** [comment]
@OfBorg build python3Packages.catboost

**Comment by pbsds** [comment]
I cherry-picked the `python3Packages.shap` fix from my other PR #187032 in hopes of actually testing this.

@ofborg build

File: pkgs/development/python-modules/catboost/default.nix [hunk 1]
```diff
@@ -1,38 +1,38 @@
 { buildPythonPackage, fetchFromGitHub, fetchpatch, lib, pythonOlder
-, clang_7, python2
+, clang_12, python, python2, sd
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:2]: Can we finally get rid of python2?

File: pkgs/development/python-modules/catboost/default.nix [hunk 2]
```diff
@@ -1,38 +1,38 @@
 { buildPythonPackage, fetchFromGitHub, fetchpatch, lib, pythonOlder
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:1]:
```suggestion
{ buildPythonPackage, fetchFromGitHub, lib, pythonOlder
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 3]
```diff
@@ -1,38 +1,38 @@
...
-    (fetchpatch {
-      name = "format.patch";
-      url = "https://github.com/catboost/catboost/pull/1528/commits/a692ba42e5c0f62e5da82b2f6fccfa77deb3419c.patch";
-      sha256 = "sha256-fNGucHxsSDFRLk3hFH7rm+zzTdDpY9/QjRs8K+AzVvo=";
-    })
   ];
 
+  postPatch = ''
+    # substituteInPlace is too slow for these large files
+    sd '\$\(YMAKE_PYTHON3-.*\)/python3' ${python}/bin/python3 make/*.makefile
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:30]: Please use sed

File: pkgs/development/python-modules/catboost/default.nix (hunk 1 above)
Comment (pbsds) [reply pkgs/development/python-modules/catboost/default.nix:2]: Most likely not, but it is only a build dep.

File: pkgs/development/python-modules/catboost/default.nix [hunk 4]
```diff
@@ -1,38 +1,38 @@
-{ buildPythonPackage, fetchFromGitHub, fetchpatch, lib, pythonOlder
-, clang_7, python2
+{ buildPythonPackage, fetchFromGitHub, lib, pythonOlder
+, clang_12, python2, python3
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:2]:
```suggestion
, clang_12, python2, python
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 5]
```diff
@@ -1,38 +1,38 @@
...
-    (fetchpatch {
-      name = "format.patch";
-      url = "https://github.com/catboost/catboost/pull/1528/commits/a692ba42e5c0f62e5da82b2f6fccfa77deb3419c.patch";
-      sha256 = "sha256-fNGucHxsSDFRLk3hFH7rm+zzTdDpY9/QjRs8K+AzVvo=";
-    })
   ];
 
+  postPatch = ''
+    # substituteInPlace is too slow for these large files, and the target has lots of numbers in it that change often.
+    sed -e 's|\$(YMAKE_PYTHON3-.*)/python3|${python3}/bin/python3|' -i make/*.makefile
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:30]:
```suggestion
    sed -e 's|\$(YMAKE_PYTHON3-.*)/python3|${python.interpreter}|' -i make/*.makefile
```
lets make sure to use the right python3.

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.catboost python310Packages.shap

**Pull request merged**